### ISSUE TITLE
BasicAuth protected Foxx API

### ISSUE THREAD

**Issue by razvanphp** [arangodb/arangodb#1290]
For the context, we developed a small Foxx app that needs authentication, and BasicAuth is enough.

Here is the current solution:
````javascript
var AUTH = 'username:password';
var Buffer = require('buffer').Buffer;

controller.before(function(req, res) {
    res.set('Content-Type', 'application/json; charset=utf-8');
    res.set('Cache-Control', 'no-store');
    res.set('Pragma', 'no-cache');
    var check = 'Basic ' + new Buffer(AUTH).toString('base64');
    if (req.headers.authorization === check) {
        return true;
    } else {
        res.status(401);
        res.set('WWW-Authenticate', 'Basic realm="Private"');
        res.json({'error': 'Unauthorized'});
        return false;
    }
});
```` 

As this endpoint will be only accessed by another system, and not real users, cookie authentication, sessions or oAuth are definitely an overkill, so maybe native BasicAuth is easy to implement as @mchacki mentioned.

The code snippet works ok when the database has the authentication turned off, but in production passwords are set, so the password for the web interface will conflict with the password the Foxx app expects, when using the 'Try Me!' function of the generated documentation.

My idea would be to have the possibility to access the Foxx app with both a valid username / password of the database itself, or to 'fallback' to a local username / password defined in the controller. This way there would be no conflict, and ideally the documentation would show status code 401 as possible. Giving out database users is also not desirable, since that would give access to all collections, not just that specific Foxx endpoint.

Thank you!

**Comment by pluma** [comment]
Foxx apps are currently covered by the database's basic auth if authentication is enabled. We're planning to change that so Foxx apps can have their own auth mechanisms even when the remaining HTTP API is restricted. See #1268 for the ongoing work for the admin app. Once the PR lands, it should be applicable to other apps, too.

**Comment by razvanphp** [comment]
I saw that PR, but it will probably land in 2.6. Is it not a simpler solution for this? The same problem is when I want to give the respective team access to the generated documentation, they must have a valid database account (since it resides in aardvark), that would allow them to read and modify all the collections..

**Comment by pluma** [comment]
Okay, sorry about the misunderstanding. I wasn't fully aware of the `authenticate-system-only` flag's behaviour.

As far as I understand it, ArangoDB shouldn't interfere if you enable that flag. If the browser sends the same credentials for the two different realms, I don't think there's much we can do to prevent that. The PR would probably help because it would let you use the API with cookie authentication (which the admin app can handle), eliminating the need for Basic auth to access the admin app.

The problem with what you're describing is that you want the documentation to be accessible to users who aren't otherwise authenticated for the admin app (which would allow them to make arbitrary modifications). As the documentation is part of the admin app rather than the Foxx app it documents, I don't see an easy way to solve that.

I'll see how difficult it would be to extract a standalone documentation app and get back to you.

**Comment by razvanphp** [comment]
Ah, my bad, you are right, it works with `--server.authenticate-system-only true`. Browsers should implement this correctly, as [RFC2617](http://tools.ietf.org/html/rfc2617) states that:

> The realm value (case-sensitive), in combination with the canonical root URL of the server being accessed, defines the protection space.

Thank you for your time regarding the documentation. I just showed it in a meeting and the logical question I got was: where can I see and test this? Then I realized I cannot give away the db password.. which defeats the purpose of the documentation somehow, as it is more useful to the user, not the developer.

**Comment by pluma** [comment]
I've written a simple documentation browser based on [swagger-ui 2.0](swagger-api/swagger-ui). It's available in the Foxx app store as "swagger" and on [GitHub](https://github.com/arangodb/foxx-swagger).

Let me know if this solves your problem.

**Comment by razvanphp** [comment]
It looks good, but probably I'm doing something wrong, because I can only get the `/_admin/aardvark` documentation. For some reason, changing the path and version in `manifest.json` has no effect. The app is in dev mode. Probably I will have to include it somehow inside my app, and not mount it in the same path, because I need common authentication.

**Comment by pluma** [comment]
You need to change the path in the app's configuration. Apparently the configuration dialog is no longer shown when mounting an application in 2.5 and later, so you need to open the app in the admin frontend's "applications" tab and then click the cog icon in the top right. It should let you enter an arbitrary mount point as "appPath".

**Comment by pluma** [comment]
Oh, and make sure to open (e.g. via the big "Open" button on the left) the actual app when you're in the admin's application detail view. I confused myself for a second because the application detail view *also* shows an API documentation ;)

**Comment by razvanphp** [comment]
Yeah, I got that, but the appPath configuration is not changing. I can see the PATCH request, but probably it's not saved, after refresh I have the admin link back there. Is it the same in 2.5.1 for you?

**Comment by pluma** [comment]
I'll look into that and get back to you as soon as I know more.

**Comment by pluma** [comment]
Apparently the PATCH endpoint expected a different format. I've adjusted it in devel and 2.5. In the meantime you can manually adjust the apps via the system collection `_apps` (just filter by `mount`, make sure to wrap the value in quotes). The fix will be in 2.5.2.

**Comment by razvanphp** [comment]
The bastard did not want with editing `_apps` collection either, he would revert it. I replicated your fix in my `foxxes.js` file and now the PATCH works. The only problem with it now is that host endpoint is `0.0.0.0:8529` so the Try me out! functionality does not work. 

Again, it looks good, kind of ruff, but it's better than nothing. I think this should be somehow integrated in any app, to be able to define a controller route with the documentation, also protected by the `.before` method password.

**Comment by pluma** [comment]
I guess we could add something like a "documentation" method to the controller (like the destroy-session method for creating logout routes). That would allow developers to expose the documentation any way they want. The bigger problem is that the assets Swagger depends on would have to be exposed somehow, too.

I'll look into the `0.0.0.0` thing. I presume your ArangoDB instance is listening globally on `8529` rather than localhost-only?

**Comment by razvanphp** [comment]
Yes, in the config I have:
> endpoint = tcp://0.0.0.0:8529

**Comment by pluma** [comment]
I've updated the swagger app to 1.0.1 and it now relies on the "Host" header. This isn't optimal in terms of security, but as it's only used to generate the "host" field in the Swagger response, I don't think this has any security implications. There's no sane way to determine what the public hostname of the app should be.

**Comment by razvanphp** [comment]
Then probably it should be a configuration field. But if you integrate this somehow, the controller should know where the endpoint is, like the admin interface does.

**Comment by razvanphp** [comment]
As I said, if you don't specify the host, it will use the same host as now, using the host header.

But still this is not working, for example:

// swagger frontend
http://domain.dev:8529/_db/bi_stage/app_follow/docs/index.html

// address where he tries to do the query
http://domain.dev:8529/app_follow/user/123

// the correct one
http://domain.dev:8529/_db/bi_stage/app_follow/user/123

**Comment by pluma** [comment]
Removed the host, added the db to the base path and updated to 1.0.2.

**Comment by razvanphp** [comment]
Works great now, Thank you! Looking forward for the integrated version :)

**Comment by razvanphp** [comment]
What about the integrated version? Do you need another ticket?
